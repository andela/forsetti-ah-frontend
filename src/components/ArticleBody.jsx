import React from 'react';
import {
  FacebookShareButton,
  TwitterShareButton,
} from 'react-share';

import {
  facebook,
  twitter,
  bookmarkWhite,
  hands,
  email
} from '../assets';

const url = 'https://forsetti-ah-frontend-staging.herokuapp.com/';

const ArticleBody = ({
  body, claps, title, shareModal
}) => (
  <div className='row article-section mt-3 '>
    <div
      className='article-body col-sm-10 col-md-10 col-lg-9
              d-flex flex-column mx-2 text-left mx-lg-5 pl-4 pb-5 offset-5'
      dangerouslySetInnerHTML={{ __html: body }}
    />
    <div className='share col-sm-2 col-md-1 col-lg-1
            mx-4 mx-md-2 mt-sm-4 mb-sm-4 my-lg-6 d-flex flex-column position-sticky px-0'
    >
      <div className='row'>
        <span className='rounded-circle social col-md-12 col-3 mb-2'>
          <a href='/' className='align-self-center'>
            <img src={hands} alt='claps' className='p-1' />
            <span className='p-0 h-50 text-dark'>
              { `${claps}` }
            </span>
          </a>
        </span>
        <span
          className='col-md-12 col-2 rounded-circle'
          onClick={shareModal}
          role='button'
          tabIndex='0'
          onKeyDown={null}
        >
          <img
            src={email}
            alt='google'
            className='py-md-2 social-icon'
          />
        </span>
        <span className='col-md-12 col-2 rounded-circle '>
          <FacebookShareButton
            quote={title}
            url={`${url}/article/${title}`}
            className='share'
          >
            <img src={facebook} alt='facebook' className='h-50 py-md-2 social-icon' />
          </FacebookShareButton>
        </span>
        <span className='col-md-12 col-2 rounded-circle '>
          <TwitterShareButton
            url={`${url}/article/${title}`}
            title={title}
            className='share'
          >
            <img src={twitter} alt='twitter' className='h-50 py-md-2 social-icon' />
          </TwitterShareButton>
        </span>
        <span className='col-md-12 col-1 rounded-circle '>
          <a href='/'>
            <img src={bookmarkWhite} alt='bookmarks' className='h-50 py-md-2 social-icon' />
          </a>
        </span>
      </div>
    </div>
  </div>
);

export default ArticleBody;
