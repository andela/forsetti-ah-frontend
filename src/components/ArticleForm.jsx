import React, { Component } from 'react';
import { connect } from 'react-redux';
import FroalaEditor from 'react-froala-wysiwyg';
import IdleTimer from 'react-idle-timer';
import { toast } from 'react-toastify';
import {
  FormGroup,
  Input,
  Label,
  Spinner
} from 'reactstrap';
import * as articleActions from '../actions';
import TagsModal from './TagsModal';
import ToastMessage from './common/ToastMessage';
import uploadImg from '../assets/images/art-image.png';

class ArticleForm extends Component {
    state = {
      title: '',
      image: null,
      imagePreview: null,
      model: '',
      tagList: '',
      hasChanges: false,
      published: false,
      imageChanged: false
    };

    componentDidMount() {
      localStorage.removeItem('slug');
    }

  onChangeHandler = ({ target: { value } }) => {
    this.setState({ title: value, hasChanges: true });
  }

  handleModelChange = (model) => {
    this.setState({ model, hasChanges: true });
  }

  imageHandler = ({ target: { files } }) => {
    this.setState({
      image: files[0],
      imagePreview: URL.createObjectURL(files[0]),
      imageChanged: true,
      hasChanges: true
    });
  }

  saveArticle = () => {
    const { props: { createArticle } } = this;
    const {
      title, hasChanges, model, tagList, image, published, imageChanged
    } = this.state;
    if (title && model && hasChanges) {
      const plainText = model.split('<p>').join('').split('</p>').join(' ');
      const description = `${plainText.slice(0, 50)}...`;
      const articleObject = {
        title,
        body: model,
        description,
        tagList: tagList || ['General'],
        image,
        currentSlug: localStorage.getItem('slug') || null,
        imageChanged
      };
      if (published) articleObject.published = published;
      this.setState({ hasChanges: false, imageChanged: false });
      createArticle(articleObject);
    }
  }

  publishHandler = (tags) => {
    this.setState({
      published: true,
      tagList: tags,
      hasChanges: true
    });
    this.saveArticle();
  }

  render() {
    const {
      model, title, imagePreview
    } = this.state;
    const { props: { article: { isLoading, saved } } } = this;
    return (
      <div className='mb-5 mt-5'>
        {saved && toast(<ToastMessage message='Saved' />, {
          type: 'success',
          closeButton: false,
          hideProgressBar: true,
          autoClose: 5000,
        })}
        <IdleTimer
          startOnMount={false}
          onIdle={this.saveArticle}
          timeout={2500}
        />
        <FormGroup>
          {isLoading && (
            <div className='text-center text-info'>
              <Spinner color='primary' />
              {' '}
            Saving...
            </div>
          )}
          <Input
            type='text'
            placeholder='Title goes here'
            value={title}
            minLength='8'
            maxLength='50'
            className='mt-3 border-0 shadow-none bold-text form-control-lg'
            onChange={this.onChangeHandler}
            required
          />
          <Label htmlFor='file' className='artlabel'>
            {' '}
            <img src={uploadImg} alt='upload' />
            <Input
              type='file'
              accept='image/*'
              name='image'
              id='file'
              onChange={this.imageHandler}
              className='artimage'
            />
          </Label>
          Upload Image
          {' '}
          {imagePreview ? <img className='mx-auto d-block .img-fluid preview' src={imagePreview} alt='preview' /> : ''}
        </FormGroup>
        <FroalaEditor
          tag='textarea'
          model={model}
          onModelChange={this.handleModelChange}
          className='border-0 shadow-none'
          config={{
            theme: 'custom',
            placeholderText: 'Tell Your Story...',
            pluginsEnabled: ['fontFamily', 'bold', 'italic', 'underline', 'fontSize', 'undo',
              'strikeThrough', 'outdent', 'indent', 'html'],
            toolbarButtons: ['fontFamily', 'bold', 'italic', 'underline', 'fontSize', '|', 'undo',
              'strikeThrough', 'outdent', 'indent', 'html'],
          }}
        />
        <TagsModal
          publishHandler={this.publishHandler}
          className='mt-50'
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  article: state.createArticle
});

const mapDispatchToProps = {
  createArticle: articleActions.createArticle
};

const ArticleFormComponent = connect(mapStateToProps, mapDispatchToProps)(ArticleForm);

export { ArticleForm, ArticleFormComponent };
