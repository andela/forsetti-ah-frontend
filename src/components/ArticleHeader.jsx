import React from 'react';
import moment from 'moment';
import { articlePlaceholder, user } from '../assets';
import Button from './common/Button';

const ArticleHeader = ({
  title,
  image,
  username,
  readingTime,
  articleImage,
  createdAt
}) => (
  <div className='container'>
    <div className='row mt-5'>
      <div className='col-md-12 text-left'>
        <h2 className='text-uppercase article-title'>
          { title }
        </h2>
      </div>
    </div>
    <div className='row mt-3 mb-4'>
      <div className='col-3 col-md-1'>
        <img
          src={image !== null || image !== undefined ? image : user}
          alt='profile'
          className='rounded-circle media m-0'
          style={{ objectFit: 'cover', width: '70px', height: '70px' }}
        />
      </div>
      <div className='col-8 col-md-11'>
        <div className='row ml-1'>
          <div className='col-md-12'>
            <p className=''>
              {username}
              <Button type='button' className='btn btn_primary btn-sm py-0 mx-2'> Follow </Button>
            </p>
          </div>
          <div className='col-md-12'>
            <span>
              {
              moment(createdAt).format('DD MMM, YYYY')
            }
            ,
              {
                ` ${Math.ceil(readingTime / 1000.0)} min read`
              }
            </span>
          </div>
        </div>
      </div>
    </div>
    <div className='row'>
      {
      articleImage === null ? <div className='pt-3' />
        : (
          <div className='col-md-12'>
            <img
              src={articleImage !== null ? articleImage : articlePlaceholder}
              alt='articleimage'
              className='col-12 col-md-11 article-image'
            />
          </div>
        )
    }
    </div>
  </div>
);

export default ArticleHeader;
