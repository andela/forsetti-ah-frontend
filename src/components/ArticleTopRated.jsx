import React from 'react';
import { withRouter } from 'react-router-dom';
import { user } from '../assets';

const ArticleTopRated = ({
  title,
  description,
  readingTime,
  image,
  firstname,
  username,
  slug,
  history
}) => (
  <div
    className='card mt-3'
    style={{ cursor: 'pointer' }}
    onClick={() => history.push(`article/${slug}`)}
    role='button'
    tabIndex='-1'
    onKeyDown={null}
  >
    <div className='card-body p-2'>
      <div className='row'>
        <span className='col' style={{ fontSize: '1.2rem' }}>
          {title}
        </span>
      </div>
      <div className='row'>
        <span className='col' dangerouslySetInnerHTML={{ __html: description }} />
      </div>
      <div className='row p-0'>
        <div className='col-8'>
          <p className='row pl-3 mt-2'>
            <img
              src={image || user}
              alt='user'
              className='rounded-circle media m-0 mr-1'
              style={{ objectFit: 'cover', width: '20px', height: '20px' }}
            />
            {
            username || firstname
         }
          </p>
        </div>
        <div className='col-4 mt-2 mb-0'>
          {
          ` ${Math.ceil(readingTime / 1000.0)} min read`
        }
        </div>
      </div>
    </div>
  </div>
);

export default withRouter(ArticleTopRated);
