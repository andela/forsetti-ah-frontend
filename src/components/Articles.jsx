import React, { Component, Fragment } from 'react';
import Button from './common/Button';
import ArticleListCard from './common/ArticleList';

class Articles extends Component {
  listArticles = () => {
    const { articles, loadMore } = this.props;
    return (
      <div className='row'>
        <div className='col-md-12'>
          {
                articles.map(article => (
                  <ArticleListCard
                    key={Math.random(Date.now())}
                    title={article.title}
                    description={article.description}
                    readingTime={article.readingTime}
                    createdAt={article.createdAt}
                    image={article.image}
                    slug={article.slug}
                  />
                ))
            }
        </div>
        <div className='col-md-12 text-center my-5'>
          <div className='row justify-content-center'>
            <div className='col-6'>
              { articles.length % 10 === 0 && (
              <Button
                className='load-btn rounded-0 font-weight-light'
                block
                onClick={loadMore}
              >
                  Load More
              </Button>
              )
              }
            </div>
          </div>
        </div>
      </div>
    );
  }

  render() {
    return (
      <Fragment>
        <div className='Articles'>
          { this.listArticles() }
        </div>
      </Fragment>
    );
  }
}
export default Articles;
