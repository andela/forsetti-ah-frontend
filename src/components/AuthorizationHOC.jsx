import React from 'react';
import { Route, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { tokenExpired } from '../utils';

const AuthorizationHOC = ({ component: Component, token, ...rest }) => (
  <Route
    {...rest}
    render={() => {
      if (token && !tokenExpired(token)) {
        return <Component />;
      }
      return <Redirect to='/' />;
    }}
  />
);

const mapStateToProps = state => ({
  token: state.auth.token,
});

const AuthorizationHOCUnit = connect(mapStateToProps)(AuthorizationHOC);

export {
  AuthorizationHOC,
  AuthorizationHOCUnit
};
