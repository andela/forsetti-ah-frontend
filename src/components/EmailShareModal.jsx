import React, { Component, Fragment } from 'react';
import {
  Modal,
  Input,
  Form,
  FormGroup,
  Label,
  FormFeedback,
  Spinner
} from 'reactstrap';
import { toast } from 'react-toastify';
import { connect } from 'react-redux';
import { cancel } from '../assets';
import {
  closeEmailShareModal,
  emailShare,
} from '../actions';
import Button from './common/Button';
import ToastMessage from './common/ToastMessage';

class EmailShareModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      invalidEmail: false,
      isDisabled: true
    };
  }

  closeModal = () => {
    const { closeModalAction } = this.props;
    closeModalAction();
  };

  handleEmail = (e) => {
    this.isValidEmail(e);
    this.setState({ [e.target.id]: e.target.value });
  };

  isValidEmail = (e) => {
    const validEmailRegx = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;
    if (!validEmailRegx.test(e.target.value)) {
      this.setState({
        invalidEmail: true,
        isDisabled: true
      });
      return;
    }
    this.setState({
      invalidEmail: false,
      isDisabled: false
    });
  };

  reset = (e) => {
    if (e.target.value.length < 1) {
      this.setState({
        invalidEmail: false,
        isDisabled: true
      });
    }
  }

  handleEmailShare = async (e) => {
    e.preventDefault();
    this.setState({
      isDisabled: true
    });
    const { email } = this.state;
    const { emailShareAction, slug, emailShareModal } = this.props;
    await emailShareAction(slug, { email });
    if (emailShareModal) {
      toast(<ToastMessage message={`Article shared successfully shared with ${email}`} />, {
        type: 'success',
        closeButton: false,
        hideProgressBar: true,
        autoClose: 3000,
      });
      this.setState({
        email: '',
        invalidEmail: false,
        isDisabled: true
      });
      this.closeModal();
    }
  }

  render() {
    const { isOpen, emailShareModal } = this.props;
    const { email, invalidEmail, isDisabled } = this.state;
    return (
      <Fragment>
        <Modal
          isOpen={isOpen}
          centered
        >
          <div className='EmailShareModal'>
            <div className='container'>
              <div className='row my-2'>
                <div className='col-md-6'>
                  <h4 className='title'>Share Article</h4>
                </div>
                <div className='col-md-6 text-right'>
                  <span
                    role='button'
                    tabIndex='0'
                    onKeyDown={null}
                    onClick={this.closeModal}
                  >
                    <img
                      src={cancel}
                      alt='cancel'
                    />
                  </span>
                </div>
              </div>
              <div className='row my-2'>
                <div className='col-md-12'>
                  <Form onSubmit={this.handleEmailShare}>
                    <div className='row'>
                      <div className='col-md-12'>
                        <FormGroup>
                          <Label className='label'>Email</Label>
                          <Input
                            type='email'
                            id='email'
                            onChange={this.handleEmail}
                            value={email}
                            invalid={invalidEmail}
                            onKeyUp={this.reset}
                          />
                          { invalidEmail && (
                          <FormFeedback>Email is invalid a simple tweak will make it valid :)</FormFeedback>
                          ) }
                        </FormGroup>
                      </div>
                      <div className='col-md-12 text-center py-3'>
                        <Button
                          type='submit'
                          className='btn submit-btn px-4'
                          disabled={isDisabled}
                        >
                          { emailShareModal.isLoading && (<Spinner size='sm' />) }
                          {' '}
                           SHARE
                        </Button>
                      </div>
                    </div>
                  </Form>
                </div>
              </div>
            </div>
          </div>
        </Modal>
      </Fragment>
    );
  }
}

export const mapStateToProps = (state) => {
  const { emailShareModal } = state;
  return { emailShareModal };
};

export const mapDispatchToProps = dispatch => ({
  closeModalAction: () => {
    dispatch(closeEmailShareModal());
  },
  emailShareAction: (slug, payload) => {
    dispatch(emailShare(slug, payload));
  }
});

const EmailShareModalComponent = connect(mapStateToProps, mapDispatchToProps)(EmailShareModal);

export {
  EmailShareModalComponent,
  EmailShareModal
};
