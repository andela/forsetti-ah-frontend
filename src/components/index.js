import Header from './common/Header';
import Footer from './common/Footer';
import VerticalListItems from './common/VerticalListItems';
import HorizontalListItems from './common/HorizontalListItems';
import Backdrop from './common/Backdrop';
import SideDrawer from './SideDrawer';
import NavBarItems from './common/NavBarItems';
import Button from './common/Button';

export {
  Header,
  Footer,
  VerticalListItems,
  HorizontalListItems,
  Backdrop,
  SideDrawer,
  NavBarItems,
  Button
};
