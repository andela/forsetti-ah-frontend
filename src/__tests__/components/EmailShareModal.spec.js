import React from 'react';
import { shallow } from 'enzyme';
import { EmailShareModal } from '../../components/EmailShareModal';

describe('Email Share Modal Component', () => {

    const props = {
        isOpen: true, 
        emailShareModal: false
    };
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<EmailShareModal {...props} />);
    })
    test('Email share modal article should render', () => {
        expect(wrapper).toMatchSnapshot();
    });
});