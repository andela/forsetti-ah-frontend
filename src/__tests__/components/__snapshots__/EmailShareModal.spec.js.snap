// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email Share Modal Component Email share modal article should render 1`] = `ShallowWrapper {}`;
